### PR TITLE
Cleanup unused parameters, vars, procedures, interfaces

### PR DESCRIPTION
**Pull request by dev-zero**
- mpi: generate interface decls based on nametype1 whenever possible
- mpi: drop unused/unexported functions, respectively export some of them (for consistency)

intended to solve parts (or maybe all) of #361

### CODE REVIEW

**Comment by dev-zero** [comment]
commit d82360c could maybe be cherry-picked into CP2K's mpi wrapper

**Comment by dev-zero** [comment]
@pseewald the commit you might interested in is `tas & tensors: drop unused (and private) functions`
@alazzaro for you starting from the `mpi: ...` commits

I took now the rather extreme approach to drop subroutines as much as possible instead of making them public. The problem of keeping dead code is that it rots (even simple functions like an iterator reset could break if people add more state to the iterator and forget to update the reset method because it is not tested), and people who start using it may not realize that they build on a possible source of error. On the other hand are we dropping code someone wrote intentionally and people will likely not grep the git history to check whether a function they need was there at some point in the past and that they could simply revive it.
The best would probably be to add some tests (especially for the utility classes), but that takes more time.
For the latter one could instead comment out the code or only leave the commented out function signature with a note, pointing people to use the VCS to recover the lost code.

**Review by pseewald — APPROVED** [review]

File: cmake/CompilerConfiguration.cmake [hunk 1]
```diff
@@ -5,7 +5,7 @@ if (CMAKE_Fortran_COMPILER_ID STREQUAL "GNU")
   endif ()
   set(CMAKE_Fortran_FLAGS_RELEASE  "-O3 -g -funroll-loops")
   set(CMAKE_Fortran_FLAGS_COVERAGE "-O0 -g --coverage -fno-omit-frame-pointer -fcheck=all -ffpe-trap=invalid,zero,overflow -fbacktrace -finit-real=snan -finit-integer=-42 -finit-derived -Werror=realloc-lhs -finline-matmul-limit=0 -Werror")
-  set(CMAKE_Fortran_FLAGS_DEBUG    "-O2 -ggdb -fno-omit-frame-pointer -fcheck=all -ffpe-trap=invalid,zero,overflow -fbacktrace -finit-real=snan -finit-integer=-42 -finit-derived -Werror=realloc-lhs -finline-matmul-limit=0 -fsanitize=undefined -fsanitize=address -fsanitize-recover=all -Warray-temporaries")
+  set(CMAKE_Fortran_FLAGS_DEBUG    "-O2 -ggdb -fno-omit-frame-pointer -fcheck=all -ffpe-trap=invalid,zero,overflow -fbacktrace -finit-real=snan -finit-integer=-42 -finit-derived -finline-matmul-limit=0 -fsanitize=undefined -fsanitize=address -fsanitize-recover=all -Wall -Wextra -Werror -Werror=realloc-lhs -Wno-error=array-temporaries -Wno-error=compare-reals -Wno-error=function-elimination -Wno-error=surprising")
```
Comment (alazzaro) [inline cmake/CompilerConfiguration.cmake:8]: More warnings, good...

**Comment by alazzaro** [comment]
Nice, very nice, it is an incredible clean-up... Only a single (minor) comment... Well done!

**Pull request merged**